internal/k8s/crdcontroller.go

```diff
@@ -208,3 +219,214 @@ func (c *AviController) SetupAKOCRDEventHandlers(numWorkers uint32) {
+	var err error
+	fqdn := hostrule.Spec.VirtualHost.Fqdn
+	foundHost, foundHR := objects.SharedCRDLister().GetFQDNToHostruleMapping(fqdn)
+	if foundHost && foundHR != hostrule.Namespace+"/"+hostrule.Name {
+		err = fmt.Errorf("duplicate fqdn %s found in %s", fqdn, foundHR)
+		status.UpdateHostRuleStatus(key, hostrule, status.UpdateCRDStatusOptions{
+			Status: lib.StatusRejected,
+			Error:  err.Error(),
+		})
+		utils.AviLog.Warnf("key: %s, msg: %v", key, err)
+		return err
+	}
+
+	refData := map[string]string{
+		hostrule.Spec.VirtualHost.WAFPolicy:                  "WafPolicy",
+		hostrule.Spec.VirtualHost.ApplicationProfile:         "AppProfile",
+		hostrule.Spec.VirtualHost.TLS.SSLKeyCertificate.Name: "SslKeyCert",
+		hostrule.Spec.VirtualHost.TLS.SSLProfile:             "SslProfile",
+		hostrule.Spec.VirtualHost.AnalyticsProfile:           "AnalyticsProfile",
+		hostrule.Spec.VirtualHost.ErrorPageProfile:           "ErrorPageProfile",
+	}
+
+	for _, policy := range hostrule.Spec.VirtualHost.HTTPPolicy.PolicySets {
+		refData[policy] = "HttpPolicySet"
+	}
+
+	for _, script := range hostrule.Spec.VirtualHost.Datascripts {
+		refData[script] = "VsDatascript"
+	}
+
+	if err := checkRefsOnController(key, refData); err != nil {
+		status.UpdateHostRuleStatus(key, hostrule, status.UpdateCRDStatusOptions{
+			Status: lib.StatusRejected,
+			Error:  err.Error(),
+		})
+		return err
+	}
+
+	status.UpdateHostRuleStatus(key, hostrule, status.UpdateCRDStatusOptions{
+		Status: lib.StatusAccepted,
+		Error:  "",
+	})
+	return nil
+}
+
+func checkRefsOnController(key string, refMap map[string]string) error {
+	for k, value := range refMap {
+		if k == "" {
+			continue
+		}
+
+		if err := checkRefOnController(key, value, k); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var refModelMap = map[string]string{
+	"SslKeyCert":             "sslkeyandcertificate",
+	"WafPolicy":              "wafpolicy",
+	"HttpPolicySet":          "httppolicyset",
+	"SslProfile":             "sslprofile",
+	"AppProfile":             "applicationprofile",
+	"AnalyticsProfile":       "analyticsprofile",
+	"ErrorPageProfile":       "errorpageprofile",
+	"VsDatascript":           "vsdatascriptset",
+	"HealthMonitor":          "healthmonitor",
+	"ApplicationPersistence": "applicationpersistenceprofile",
+	"ServiceEngineGroup":     "serviceenginegroup",
+	"Network":                "network",
+}
+
+// checkRefOnController checks whether a provided ref on the controller
+func checkRefOnController(key, refKey, refValue string) error {
+	uri := fmt.Sprintf("/api/%s?name=%s&fields=name,type,labels,created_by", refModelMap[refKey], refValue)
+	clients := avicache.SharedAVIClients()
+
+	// assign the last avi client for ref checks
+	aviClientLen := lib.GetshardSize()
+	result, err := lib.AviGetCollectionRaw(clients.AviClient[aviClientLen], uri)
+	if err != nil {
+		utils.AviLog.Warnf("key: %s, msg: Get uri %v returned err %v", key, uri, err)
```

Comment:
If we are returning an error below, there's no need to print the warning here, right?